featurebyte/utils/async_helper.py

```diff
@@ -5,18 +5,137 @@
 from __future__ import annotations
 
 import asyncio
+import time
+import uuid
 from asyncio import Future
 from typing import Any, Coroutine, List
 
+from redis import Redis
 
-def asyncio_gather(*coros: Coroutine[Any, Any, Any], max_concurrency: int = 0) -> Future[List[Any]]:
+
+class RedisFairCountingSemaphore:
+    """
+    Fair counting semaphore using Redis adapted from "Redis In Action" by Josiah L. Carlson
+    """
+
+    def __init__(self, redis: Redis[Any], key: str, limit: int, timeout: int = 10):
+        """
+        Initialize semaphore
+
+        Parameters
+        ----------
+        redis: Redis[Any]
+            Redis connection
+        key: str
+            Semaphore key
+        limit: int
+            Semaphore limit count
+        timeout: int
+            Timeout in seconds
+        """
+        self._redis = redis
+        self._key = key
+        self._limit = limit
+        self._timeout = timeout
+        self._czset = f"{self._key}:owner"
+        self._ctr = f"{self._key}:counter"
+        self._lock = f"{self._key}:lock"
+        self._identifier: str | None = None
+
+    async def __aenter__(self) -> None:
+        """
+        Acquire semaphore on enter, blocks until semaphore is acquired or timeout
+
+        Raises
+        ------
+        TimeoutError
+            If semaphore is not acquired within timeout
+        """
+        # Acquire lock for semaphore
+        lock_timeout = 1
+        lock = self._redis.lock(self._lock, timeout=lock_timeout)
+        now = time.time()
+        while not lock.acquire(blocking=False):
+            if time.time() - now > self._timeout:
+                raise TimeoutError(f"Failed to acquire lock: {time.time() - now}")
+            await asyncio.sleep(0.1)
```

Comment:
Is the main purpose of not using `blocking_timeout` in `Redis.lock` is because of the block there is not async? Would be good add a documentation here so that we won't modify this part later & break something unintentionally.

Reply:
Sure, will add a comment. Generally we should not be using non async blocking calls in async functions as that will block other threads from execution